pkg/spire/controller.go

```diff
@@ -0,0 +1,276 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spire
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	entryv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/server/entry/v1"
+	spiffetypes "github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/spiffe/go-spiffe/v2/spiffetls/tlsconfig"
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	spireconfig "github.com/tektoncd/pipeline/pkg/spire/config"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+)
+
+type spireControllerAPIClient struct {
+	config       spireconfig.SpireConfig
+	serverConn   *grpc.ClientConn
+	workloadConn *workloadapi.X509Source
+	entryClient  entryv1.EntryClient
+	workloadAPI  *workloadapi.Client
+}
+
+func (sc *spireControllerAPIClient) checkClient(ctx context.Context) error {
+	if sc.entryClient == nil || sc.workloadConn == nil || sc.workloadAPI == nil || sc.serverConn == nil {
+		return sc.dial(ctx)
+	}
+	return nil
+}
+
+func (sc *spireControllerAPIClient) dial(ctx context.Context) error {
+	if sc.workloadConn == nil {
+		// Create X509Source
+		source, err := workloadapi.NewX509Source(ctx, workloadapi.WithClientOptions(workloadapi.WithAddr("unix://"+sc.config.SocketPath)))
+		if err != nil {
+			return fmt.Errorf("unable to create X509Source for SPIFFE client: %w", err)
+		}
+		sc.workloadConn = source
+	}
+
+	if sc.workloadAPI == nil {
+		client, err := workloadapi.New(ctx, workloadapi.WithAddr("unix://"+sc.config.SocketPath))
+		if err != nil {
+			return fmt.Errorf("spire workload API not initialized due to error: %w", err)
+		}
+		sc.workloadAPI = client
+	}
+
+	if sc.serverConn == nil {
+		// Create connection
+		tlsConfig := tlsconfig.MTLSClientConfig(sc.workloadConn, sc.workloadConn, tlsconfig.AuthorizeAny())
+		conn, err := grpc.DialContext(ctx, sc.config.ServerAddr, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+		if err != nil {
+			sc.workloadConn.Close()
+			sc.workloadConn = nil
+			return fmt.Errorf("unable to dial SPIRE server: %w", err)
+		}
+		sc.serverConn = conn
+	}
+
+	if sc.entryClient == nil {
+		sc.entryClient = entryv1.NewEntryClient(sc.serverConn)
+	}
+
+	return nil
+}
+
+// NewSpireControllerAPIClient creates a new NewSpireControllerAPIClient for the pipeline controller
+func NewSpireControllerAPIClient(c spireconfig.SpireConfig) ControllerAPIClient {
+	if c.MockSpire {
+		return &MockClient{}
+	}
+	return &spireControllerAPIClient{
+		config: c,
+	}
+}
```

Comment:
I would prefer to have two different factories here, one the tests and one for the actual usage, instead of embedding a `MockSpire` attribute in the client struct `spireControllerAPIClient`, it would be more consistent with all the other clients used in the controllers, like Tekton API clients, cloudevents client and cache client.

For instance, if you look at the https://github.com/tektoncd/pipeline/tree/main/pkg/reconciler/events/cache package, there are two clients defined there, the [regular one](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/events/cache/cache.go) and the [mock one](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/events/cache/cachefakeclient.go). 

The `knative/pkg` injection machinery makes sure that the correct client (regular or mock) is injected into the context, so that when you [get the client from the context](https://github.com/tektoncd/pipeline/blob/65d3bf3b2d1a828c4598ede08416aab899e2d779/pkg/reconciler/events/cloudevent/cloud_event_controller.go#L143) you get a different one, depending on whether it's unit tests or actual running code.

Reply:
@mattmoor may want to have a say on this too 🙏

Reply:
Interesting. I did not know about this. I will update to use that.

Reply:
Hmm I see. I might need some help setting this up. I am not sure how this would work in this context.

Reply:
seems do-able, does this look about right? https://github.com/pxp928/pipeline/pull/26/files

Reply:
Thanks @lumjjb - that's definitely in the right direction! 

The only question I have is whether we could get the config from the context, so that when the client are setup the config is added directly, and we don't need to add it later on. Ideally we should add the client config once, unless we need to support reloading configuration changes at run time. Wdyt?

Reply:
Thanks @afrittoli ! It's in a weird spot with the [initialization of the TaskRun reconciler](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/taskrun/controller.go#L47), since some of the options come in as CLI arguments. Do you have any recommendations to handle this?